_posts/2017-05-04-docker-swarm-part-one.md

```diff
@@ -0,0 +1,255 @@
+systemctl daemon-reload
+systemctl restart docker
+
+apt-get install -y -q tree
+```
+
+## Connect workers and managers
+I am going to assume that you have docker installed on your machine in order to run commands in the nodes. Also, you can ssh one of the VMs and run docker client from there, however, I recommend to run from your local machine so you can have easy access to your files.
+
+### Run commands from the manager
+
+Export Docker host variable with manager address:
+export DOCKER_HOST=192.168.99.201
+
+Initialize the swarm manager node:
+
+`docker swarm init --advertise-addr 192.168.99.201`
+
+The output of this command contains the command you must run in the nodes to join them as workers:
+
+```bash
+docker swarm join \
+--token SWMTKN-1-5115o3dmke8531usz4f4vqtqj9u25j7lflh659oud9kt8qboez-cdqnqq7f3mlvx67gkk2x65pwj \
+192.168.99.201:2377
+```
+
+The **SWarM ToKeN** version **1** corresponds to the pattern `SWMTKN-1-< digest-of-root-CA-cert>-< random-secret >`. [Here](https://github.com/docker/labs/tree/master/security/swarm#step-2-add-a-new-manager) you can find detailed info.
+
+You can get the command to join as worker or manager running `docker swarm join-token worker` or `docker swarm join-token worker` in any manager.
+
+After joining the nodes you can see the status. To achieve that export a manager address `export DOCKER_HOST=192.168.99.201` of the cluster running `[docker node ls](https://docs.docker.com/engine/swarm/manage-nodes/#list-nodes)`.
+
+Example:
+
+```
+$ docker node ls
+ID                           HOST  NAME  STATUS  AVAILABILITY  MANAGER STATUS
+67amhouicfyqunzs6k5p3g6cj *  m2        Ready   Active        Reachable
+du1d5khn10hp1ut3v3134ll5m    w1        Ready   Active
+lvj934al8rozerkql3mjlwink    m1        Ready   Active        Leader
+ywhwdja3u2csui3ha9ghsmwqc    w2        Ready   Active
+```
+
+To see the swarm status of a specific node use `node info`
+
+```
+$ docker info | grep --after-context=4 Swarm
+Swarm: active
+ NodeID: 67amhouicfyqunzs6k5p3g6cj
+ Is Manager: true
+ ClusterID: 1mpireicaca699772x3agxyca
+ Managers: 2
+ Nodes: 4
+```
+
+ For any of the status commands you can always run watch in another console and see how the things change. For instance, you can run `watch -d -n 1 docker node ls` before adding new nodes and see how it changes. The command flashes the differences (-d) and runs the command each second (-n 1).
+
+Challenge: Try to connect the m2 instance as a manager.
+
+### Create the stack
+We are going to create services using `docker stack`. This is a similar to docker-compose but for services. You can always do the same with `docker service`.
+
+In this post, we are going to create only a service and prove that we have a load balancer behind. We need to create a file that represents the definition of the services:
+
+swarmrt.yml
+
+```
+version: '3.1'
+
+services:
+    swarmrt:
+        image: charlieraffellini/swarmrt
```

Comment:
It would be good to explain that this image must be created before on each node (or maybe I'm wrong).